src/mono/mono/mini/aot-compiler.c

```diff
@@ -13142,77 +13211,166 @@ add_profile_instances (MonoAotCompile *acfg, ProfileData *data)
-				//printf ("%s\n", mono_method_full_name (m, TRUE));
-				add_profile_method (acfg, m);
-				count ++;
-			} else {
-				//printf ("SKIP: %s (%s)\n", mono_method_get_full_name (m), acfg->image->name);
-			}
-			/*
-			 * FIXME: We might skip some instances, for example:
-			 * Foo<Bar> won't be compiled when compiling Foo's assembly since it doesn't match the first case,
-			 * and it won't be compiled when compiling Bar's assembly if Foo's assembly is not loaded.
-			 */
-		}
+		count += add_single_profile_method (acfg, methodEntry);
+	}
+	return count;
+}
+
+//---------------------------------------------------------------------------------------
+//
+// add_mibc_profile_methods is the overarching method that adds methods within a .mibc
+// profile file to be compiled ahead of time. .mibc is a portable executable with
+// methods grouped under mibcGroupMethods, which are summarized within the global
+// function AssemblyDictionary. This method obtains the AssemblyDictionary and iterates
+// over il opcodes and arguments to retrieve mibcGroupMethods and thereafter calls
+// add_mibc_group_method_methods.
+//
+// Sample AssemblyDictionary format
+//
+// Method 'AssemblyDictionary' (#2f67) (0x06000006)
+// {
+//   IL_0000:  ldstr      0x70000001
+//   IL_0005:  ldtoken    0x06000001 // mibcGroupMethod
+//   IL_000a:  pop
+//   ...
+// }
+//
+// Arguments:
+// 	* acfg - MonoAotCompiler instance
+//	* filename - the .mibc profile file containing the methods to AOT compile
+//
+
+static void
+add_mibc_profile_methods (MonoAotCompile *acfg, char *filename)
+{
+	MonoImageOpenStatus status = MONO_IMAGE_OK;
+	MonoImage *image = mono_image_open_a_lot (mono_alc_get_default (), filename, &status, TRUE);
+	g_assert (image != NULL);
+	g_assert (status == MONO_IMAGE_OK);
+
+	ERROR_DECL (error);
+
+	MonoClass *mibcModuleClass = mono_class_from_name_checked (image, "", "<Module>", error);
+	mono_error_assert_ok (error);
+
+	MonoMethod *assemblyDictionary = mono_class_get_method_from_name_checked (mibcModuleClass, "AssemblyDictionary", 0, 0, error);
+	MonoGenericContext *context = mono_method_get_context (assemblyDictionary);
+	mono_error_assert_ok (error);
+
+	MonoMethodHeader *header = mono_method_get_header_internal (assemblyDictionary, error);
+	mono_error_assert_ok (error);
+
+	int count = 0;
+	uint8_t *cur = (uint8_t*)header->code;
+	uint8_t *end = (uint8_t*)header->code + header->code_size;
+	while (cur < end) {
+		MonoOpcodeEnum il_op;
+		const unsigned char *opcodeIp = (unsigned char*)cur;
+		const unsigned char *opcodeEnd = (unsigned char*)end;
+		cur += mono_opcode_value_and_size (&opcodeIp, opcodeEnd, &il_op);
+		// opcodeIp gets moved to point at end of opcode
+		// il opcode arg is opcodeIp + 1
+		// we only care about args of ldtoken's, which are 32bits/4bytes
+		if (il_op != MONO_CEE_LDTOKEN)
+			continue;
+
+		g_assert (opcodeIp + 4 < opcodeEnd);
+		guint32 token = *(guint32 *)(opcodeIp + 1);
```

Comment:
```suggestion
		guint32 token = read32 (opcodeIp + 1);
```

Reply:
Thanks! I thought it would work, but convinced myself it didnt when xcode lldb complained about `expr read32 (opcodeIp + 1)`. Is there a way to have the xcode lldb recognize methods such as `read32`?

Reply:
I think if you compile the runtime with "enough" debug information (maybe `-g3`) then debuggers ought to recognize macros like `read32`.  not sure if that really works with clang+lldb